python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -699,6 +700,131 @@ def ngroup(self, ascending=True):
+                "Sorry, only integer seeds are supported for random_state "
+                "in this case"
+            )
+        # Get the groups
+        try:
+            _, (index,), group_offsets = self._groupby.groups(
+                [*self.obj._index._columns]  # type: ignore
+            )
+        except ValueError:
+            raise NotImplementedError(
+                "Sorry groupby.sample with multiindex not implemented"
+            )
+        group_offsets = np.asarray(group_offsets).astype(np.int32)
+        size_per_group = np.diff(group_offsets)
+        if n is not None:
+            samples_per_group = np.broadcast_to(
+                np.int32(n), size_per_group.shape
+            )
+            if not replace and (minsize := size_per_group.min()) < n:
+                raise ValueError(
+                    f"Cannot sample {n=} without replacement, "
+                    f"smallest group is {minsize}"
+                )
+        else:
+            # Pandas uses round-to-nearest, ties to even to
+            # pick sample sizes for the fractional case (unlike IEEE
+            # which is round-to-nearest, ties to sgn(x) * inf).
+            samples_per_group = np.round(
+                size_per_group * frac, decimals=0
+            ).astype(np.int32)
+        if replace:
+            # Use numpy to do all-at-once generation of the
+            # We can use numpy in this case, which is fast for lots of
+            # groups and large samples
+            lo = 0
+            hi = np.repeat(size_per_group, samples_per_group)
+            rng = np.random.default_rng(seed=random_state)
+            # Would be nice to use cupy here, but their rng.integers
+            # interface doesn't take array lo and hi arguments.
+            indices = rng.integers(lo, hi, dtype=np.int32).reshape(-1)
+            indices += np.repeat(group_offsets[:-1], samples_per_group)
+        else:
+            # Approach: do a segmented argsort of the index array and take
+            # the first samples_per_group entries from sorted array.
+            if len(size_per_group) < 500:
+                # Empirically shuffling with cupy is faster at this scale
+                indices = cp.asarray(index.data_array_view(mode="read"))
+                rs = cp.random.get_random_state()
+                rs.seed(seed=random_state)
+                for off, size in zip(group_offsets, size_per_group):
+                    rs.shuffle(indices[off : off + size])
+            else:
+                rng = cp.random.default_rng(seed=random_state)
+                offsets_col = as_column(group_offsets)
+                values = [index]
+                keys = [as_column(rng.random(size=len(index)))]
+                (indices,) = segmented_sort_by_key(
+                    values, keys, offsets_col, [], []
+                )
+                indices = cp.asarray(indices.data_array_view(mode="read"))
+            # For many groups this is the slowest part.
```

Comment:
Is the slow part because this is numpy? The slow bit is generating “to_take” and not the actual taking, right? I have a solution I can write up in a bit if that’s accurate